File: 2022/Day 03/pt2.cpp
#include "helper.hpp"

int get_group_rucksack_priority(std::string l0, std::string l1, std::string l2)
{
    std::vector<char> l0_vector, l1_vector, l2_vector;
    std::copy(l0.begin(), l0.end(), std::back_inserter(l0_vector));
    std::copy(l1.begin(), l1.end(), std::back_inserter(l1_vector));
    std::copy(l2.begin(), l2.end(), std::back_inserter(l2_vector));

    std::vector<char> l01_intersect = Helper::intersect_vectors(l0_vector, l1_vector);
    std::vector<char> l012_intersect = Helper::intersect_vectors(l01_intersect, l2_vector);

    char priority = l012_intersect[0];
    if (priority > 'a')
        return priority - 'a' + 1;
    return priority - 'A' + 27;
}

int get_priority_sum(std::vector<std::string> lines)
{
    int priority = 0;

    for (int i = 0; i < lines.size(); i += 3)
    {
        priority += get_group_rucksack_priority(lines[i], lines[i + 1], lines[i + 2]);
    }

    return priority;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");

    auto t1 = Helper::get_current_time();
    int output = get_priority_sum(lines);
    std::cout << "Output: " << output << std::endl;
    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/day_07/pt2.go
package day07

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/benjababe/advent-of-code/helper"
)

var cardOrder2 = "J23456789TQKA"
var jCountMap = map[string][]int64{
	"highCard":     {1, 2, 4, 6, 7, 7},
	"onePair":      {2, 4, 6, 7, 7, 7},
	"twoPair":      {3, 5, 7, 7, 7, 7},
	"threeOfAKind": {4, 6, 7, 7, 7, 7},
	"fullHouse":    {5, 5, 5, 5, 5, 5},
	"fourOfAKind":  {6, 7, 7, 7, 7, 7},
	"fiveOfAKind":  {7, 7, 7, 7, 7, 7},
}

func getHandScore2(hand string) int64 {
	hScore := int64(0)
	comboCounts := [6]int{0, 0, 0, 0, 0, 0}

	sortedHand := helper.SortString(hand)

	comboType := ""
	curChar, curCount, jCount := '0', 0, int64(0)
	sortedHand += " "

	for _, char := range sortedHand {
		switch char {
		case 'J':
			jCount++
		case curChar:
			curCount++
		default:
			comboCounts[curCount]++
			curChar = char
			curCount = 1
		}
	}

	switch {
	case comboCounts[2] == 1 && comboCounts[3] == 0:
		comboType = "onePair"
	case comboCounts[2] == 2:
		comboType = "twoPair"
	case comboCounts[2] == 0 && comboCounts[3] == 1:
		comboType = "threeOfAKind"
	case comboCounts[2] == 1 && comboCounts[3] == 1:
		comboType = "fullHouse"
	case comboCounts[4] == 1:
		comboType = "fourOfAKind"
	case comboCounts[5] == 1:
		comboType = "fiveOfAKind"
	default:
		comboType = "highCard"
	}

	hScore = jCountMap[comboType][jCount]
	return hScore
}

func solvePt2(lines []string) int64 {
	score := int64(0)
	hands := []Hand{}

	for _, line := range lines {
		lineSpl := strings.Split(line, " ")
		handStr, bidStr := lineSpl[0], lineSpl[1]

		bid, _ := strconv.ParseInt(bidStr, 10, 64)
		handScore := getHandScore2(handStr)

		hands = append(hands, Hand{bid, handScore, handStr})
	}

	sortHands(hands, cardOrder2)
	for i, hand := range hands {
		score += int64(i+1) * hand.bid
	}

	return score
}

func Pt2() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solvePt2(lines)
	fmt.Printf("Day 7\tPt2:\t%d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)
}
